providers/upgrade-zfsLocalPV/test.yml

```diff
@@ -0,0 +1,209 @@
+
+     - name: Record the number of zfs-controller replicas
+       set_fact:
+         zfs_ctrl_replicas: "{{ no_of_zfs_ctrl_replicas.stdout }}"
+
+     - name: Get the list of node-agent pods in openebs-zfs-node daemonset
+       shell: >
+         kubectl get po -n kube-system -l app=openebs-zfs-node --no-headers -o custom-columns=:.metadata.name
+       args:
+         executable: /bin/bash
+       register: ds_pods
+         
+     - block:
+       ## This task creates new CRDs as zfs-LocalPV related CRs are now grouped 
+       ## under `zfs.openebs.io` from v0.6 release.
+       - name: Apply the new CRDs for zfs-LocalPV 
+         shell: >
+           kubectl apply -f https://raw.githubusercontent.com/openebs/zfs-localpv/master/upgrade/crd.yaml
+         args:
+           executable: /bin/bash
+         register: new_crds
+         failed_when: "new_crds.rc != 0"
+
+       ## This task create new CRs for zfs-volume and zfs-snapshot with updated
+       ## apiversion to `zfs.openebs.io`. Previously this was `openebs.io`.
+       - name: Download the Upgrade script for creating new CRs with apiversion as `zfs.openebs.io`
+         get_url:  
+           url: https://raw.githubusercontent.com/openebs/zfs-localpv/master/upgrade/upgrade.sh
+           dest: ./upgrade.sh
+           force: yes
+         register: result
+         until: "'OK' in result.msg"
+         delay: 5
+         retries: 3
+
+       - name: Apply the upgrade scriptopenebs-zfs-node
+         shell: sh ./upgrade.sh {{ operator_ns }}
+         args:
+           executable: /bin/bash
+
+       when: 
+       - zfs_driver_tag.stdout == "v0.4" or zfs_driver_tag.stdout == "0.4.1" or zfs_driver_tag.stdout == "v0.5"
+       - "'v0.4.x' not in to_version_zfs_branch" 
+       - "'v0.5.x' not in to_version_zfs_branch"
+
+     - name: Download the zfs-operator file
+       get_url:
+         url: https://raw.githubusercontent.com/openebs/zfs-localpv/{{ to_version_zfs_branch }}/deploy/zfs-operator.yaml
+         dest: ./new_zfs_operator.yml
+         force: yes
+       register: result
+       until: "'OK' in result.msg"
+       delay: 5
+       retries: 3
+            
+     - name: Update the openebs zfs-driver image 
+       replace:
+         path: ./new_zfs_operator.yml
+         regexp: quay.io/openebs/zfs-driver:ci
+         replace: "{{ lookup('env','ZFS_DRIVER_IMAGE') }}"
+       when: lookup('env','ZFS_DRIVER_IMAGE') | length > 0
+
+     - name: Update the number of zfs-controller statefulset replicas
+       replace:
+         path: ./new_zfs_operator.yml
+         regexp: "replicas: 1"
+         replace: "replicas: {{ zfs_ctrl_replicas }}"
+
+     - name: Apply the zfs_operator file to deploy zfs-driver components to the newer version
+       shell: 
+         kubectl apply -f ./new_zfs_operator.yml
+       args:
+          executable: /bin/bash
+
+     - name: Wait for some time to old zfs-driver components to go into Terminating state.
+       shell: >
```

Comment:
this may not be sufficient in all cases. Better to include a condition to check if the pod is no more

Reply:
As zfs-controller is a statefulset pod so pod name will be same all the time `openebs-zfs-controller-0` like that. For node-agent daemonset i'm checking that old pods are not present.
Waiting time is not for confirming that pod are terminated. It is just to avoid the check condition for pods where all pods in running state at very initial.
And later we are checking the desired replicas and upgraded image.